src/server/index.js

```diff
@@ -217,6 +218,8 @@ async function NextAuth (req, res, userSuppliedOptions) {
       redirect
     }
 
+    const pkce = options.providers[provider]?.pkce ? pkceChallenge(64) : undefined
```

Comment:
length of `64` was arbitrarily chosen here. The [requirement by the spec](https://tools.ietf.org/html/rfc7636#section-4.1) is 43-128 characters

Reply:
In terms of saving the challenge and verifier, it could be saved in local storage (maybe session) to be cleaned up after a successful log in.

Reply:
I am not sure we would like to store anything sensitive this way, cause it is perceptible with potentially malicious code on the client. If I understand it correctly, `sessionStorage` would be a better option in this cause, although I am still not sure it is good enough. Going to discuss it with @iaincollins hopefully this weekend, I hope he has some insights.

Reply:
No problem! I inspired myself from [oidc-client-js](https://github.com/IdentityModel/oidc-client-js) which I'm currently using in a production spa. They use `localStorage` for the code_verifier and `sessionStorage` for the tokens. 

Via the spec every new authorization flow should use a new `code_verifier` so this could be why they've opted to go this route. I'm not really sure either what kind of security concerns this method raises.

In any case I've set up a simple sample to generate the `code_verifier` and `code_chalenge` in the browser. I don't think it's perfect but it can help to get you started. https://gist.github.com/mikecabana/9cae8b447afd2a36da5d38b94bfc2565

Reply:
In theory we wouldn't actually need PKCE to be secure I think, since our secrets are all server-side and never accessed by the client. So this is why I went with my implementation. Although there might be something important I leave out. We will have a talk about this soon.

Reply:
@balazsorban44 FWIW for the record this is exactly my current understanding as well.